Add lint rule against .tooltipped and introduce accessibility.yml config

khiga8 opened the pull request:
### What are you trying to accomplish?

I started this rule in dotcom, but then I realized that PVC might make more sense so this rule can be shared across many projects.

This PR:
* Introduce a lint rule against `.tooltipped` in favor of alternatives, or the more recent tooltip.
* Started a `accessibility.yml` which will allow consumers to easily pull in recommended accessibility linting configurations which we can count against the scorecard. This will specifically contain rules that count towards the accessibility scorecard.
* I moved helpers out of `base_linter.rb` to a helpers module. `BaseLinter` seems to come with a lot of additional things this rule doesn't need, but I need to use these helpers. 

I _think_ consumers can now add the following in their `.erb-lint.yml` to automatically pull in rules:

```yaml
inherit_gem:
  primer_view_components:
    - primer/view_components/linters/accessibility.yml
```

### Integration
No

#### List the issues that this change affects.

Closes https://github.com/github/accessibility/issues/3704

#### Risk Assessment

- [x] **Low risk** the change is small, highly observable, and easily rolled back.
- [ ] **Medium risk** changes that are isolated, reduced in scope or could impact few users. The change will not impact library availability.
- [ ] **High risk** changes are those that could impact customers and SLOs, low or no test coverage, low observability, or slow to rollback.

### What approach did you choose and why?

* I introduced a `accessibility.yml` config so consumers can easily configure recommended rules.
* I namespaced the rule as `Primer::Accessibility`.

### Anything you want to highlight for special attention from reviewers?

I namespaced the rule `Primer::Accessibility::` to align with what we do in [erblint-github](https://github.com/github/erblint-github). 

The pro of this is that it can help to identify where a rule originates from. The con, is that the disable comment is kind of long (e.g. `erblint:disable Primer::Accessibility::TooltipHasMigration).

This also isn't consistent with the other rules living in PVC. What do you think? @primer/rails-reviewers?? 📣

I would also like to add docs for this rule but I'm not sure where it belongs, also given the doc migration work. What do you think? 📣

### Merge checklist

- [ ] Added/updated tests
- [ ] Added/updated documentation
- [ ] Added/updated previews (Lookbook)
- [ ] Tested in Chrome
- [ ] Tested in Firefox
- [ ] Tested in Safari
- [ ] Tested in Edge

Take a look at the [What we look for in reviews](https://github.com/primer/react/blob/main/contributor-docs/CONTRIBUTING.md#what-we-look-for-in-reviews) section of the contributing guidelines for more information on how we review PRs.

khiga8 commented on lib/primer/view_components/linters/accessibility.yml:
```diff
@@ -0,0 +1,7 @@
+# Accessibility-focused lint rule configurations, tracked by accessibility scorecard.
```
Should I move this somewhere else?

jonrohan replied:
This seems fine

camertron reviewed (approved):
This looks good to me 👍 Have you tested that it works without choking in dotcom?

khiga8 commented:
> This looks good to me 👍 Have you tested that it works without choking in dotcom?

I'll give it a try! And also fix the tests, which are breaking for some reason :(

khiga8 commented on lib/primer/view_components/linters/tooltipped_migration.rb:
```diff
@@ -0,0 +1,37 @@
+          MIGRATE_TO_NEWER_TOOLTIP = ".tooltipped has been deprecated. Tooltip should never be set on a non-interactive element (e.g. <span>, <div>, <li>) " \
+            "Find alternatives in https://primer.style/design/guides/accessibility/tooltip-alternatives/." \
+            "If you're setting a tooltip on an interactive element, use the latest tooltip in https://primer.style/view-components."
+          TOOLTIPPED_RUBY_PATTERN = /classes:.*tooltipped/.freeze
```
TIL

khiga8 replied:
Done in https://github.com/primer/view_components/pull/2086/commits/09edb8dffacc5d22f394c15bb62dbf2da233c5b9 ✅

khiga8 commented on test/lib/erblint/tooltipped_migration_test.rb:
```diff
@@ -0,0 +1,37 @@
+
+  def test_does_not_warn_if_inline_disable_comment
+    @file = <<~HTML
+      <%= render SomeComponent.new(classes: "tooltipped") do %><%# erblint:disable Primer::Accessibility::TooltippedMigration %>
```
Introducing this new ERBLint rule with `Primer::` resulted in a [bunch of tests breaking](https://github.com/primer/view_components/actions/runs/5326286252/jobs/9648126723) due to namespace confusion:

```rb
NameError: uninitialized constant ERBLint::Linters::Primer::Classify
```

Wherever we call the helper `Primer::Classify`, it started to try to call `ERBLint::Linters::Primer::Classify`! 

I resolved this by prefixing these calls with `::` in https://github.com/primer/view_components/pull/2086/commits/8129466bd5452917f20afc737d9f8fb58820dc69 and https://github.com/primer/view_components/pull/2086/commits/43e9e500daa085bc894cf696dd6aac424426e048.  (Source: [Stackoverflow](https://stackoverflow.com/a/29920332)


This is a new Ruby syntax for me so I'd appreciate eyes! (cc: @camertron)

camertron replied:
Nice yeah that should be totally fine 👍

khiga8 commented:
> Have you tested that it works without choking in dotcom?

@camertron I started a test branch using the vendor script and [pointing to this commit in dotcom](https://github.com/github/github/compare/kh-run-erblint?expand=1). I updated the `.erb-lint.yml` config and got this rule running, and tried inserting inline disables. 💪

The one unexpected thing was the path which I had to set to get this config to work: [Related commit](https://github.com/github/github/commit/924d68cbce9e9c4c196e701625e7b6402f05e788#diff-09d428dfddfee8285e0a4c19ae57822ebb38b78c405a75e7f63b721f6fb00f7fR7). I did not expect to need to prefix the path to the config with `lib`. I didn't need to do that with `erblint-github`. I wonder why? It's not a blocker though.

khiga8 commented:
Since this was last reviewed, I made these updates @primer/rails-reviewers:

* Updated logic to include `class` as suggested in https://github.com/primer/view_components/pull/2086#discussion_r1237139807
* Fixed breaking tests by prefixing `::` to some of the Primer helper calls.
* Enabled this config for this project https://github.com/primer/view_components/pull/2086/commits/aa180043aa5f789edcaad234c07180e5e0acd29b and added tests to make sure the config pulls in rules as expected.

Your input is appreciated here:

* https://github.com/primer/view_components/pull/2086#discussion_r1238663883
* https://github.com/primer/view_components/pull/2086#issuecomment-1602853655

Thank you so much 🙏.

jonrohan reviewed (approved):
Looks good including the `::` prefix 👍🏻
